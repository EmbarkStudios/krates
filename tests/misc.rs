mod util;

#[test]
fn iter_names() {
    let contents = std::fs::read_to_string("tests/all-features.json")
        .map_err(|e| format!("failed to load metadata file: {}", e))
        .unwrap();

    let md: krates::cm::Metadata = serde_json::from_str(&contents)
        .map_err(|e| format!("failed to deserialize metadata: {}", e))
        .unwrap();

    let krates: krates::Krates = krates::Builder::new()
        .build_with_metadata(md, |_| {})
        .unwrap();

    let mut iter = krates.krates_by_name("winapi");

    let win28 = iter.next().unwrap();
    assert_eq!(win28.krate.name, "winapi");
    assert_eq!(
        win28.krate.version,
        krates::semver::Version::parse("0.2.8").unwrap()
    );

    let win38 = iter.next().unwrap();
    assert_eq!(win38.krate.name, "winapi");
    assert_eq!(
        win38.krate.version,
        krates::semver::Version::parse("0.3.9").unwrap()
    );

    assert!(iter.next().is_none());

    let mut iter = krates.krates_by_name("a");

    let a = iter.next().unwrap();
    assert_eq!(a.krate.name, "a");
    assert_eq!(
        a.krate.version,
        krates::semver::Version::parse("0.1.0").unwrap()
    );

    assert!(iter.next().is_none());

    let mut iter = krates.krates_by_name("winapi-x86_64-pc-windows-gnu");

    let wingnu = iter.next().unwrap();
    assert_eq!(wingnu.krate.name, "winapi-x86_64-pc-windows-gnu");
    assert_eq!(
        wingnu.krate.version,
        krates::semver::Version::parse("0.4.0").unwrap()
    );

    assert!(iter.next().is_none());
}

#[test]
fn iter_matches() {
    let contents = std::fs::read_to_string("tests/all-features.json")
        .map_err(|e| format!("failed to load metadata file: {}", e))
        .unwrap();

    let md: krates::cm::Metadata = serde_json::from_str(&contents)
        .map_err(|e| format!("failed to deserialize metadata: {}", e))
        .unwrap();

    let krates: krates::Krates = krates::Builder::new()
        .build_with_metadata(md, |_| {})
        .unwrap();

    {
        let any = krates::semver::VersionReq::STAR;
        let mut iter = krates.search_matches("winapi", any);

        let win28 = iter.next().unwrap();
        assert_eq!(win28.krate.name, "winapi");
        assert_eq!(
            win28.krate.version,
            krates::semver::Version::parse("0.2.8").unwrap()
        );

        let win38 = iter.next().unwrap();
        assert_eq!(win38.krate.name, "winapi");
        assert_eq!(
            win38.krate.version,
            krates::semver::Version::parse("0.3.9").unwrap()
        );

        assert!(iter.next().is_none());
    }

    {
        let two = krates::semver::VersionReq::parse("=0.2").unwrap();
        let mut iter = krates.search_matches("winapi", two);

        let win28 = iter.next().unwrap();
        assert_eq!(win28.krate.name, "winapi");
        assert_eq!(
            win28.krate.version,
            krates::semver::Version::parse("0.2.8").unwrap()
        );

        assert!(iter.next().is_none());
    }

    {
        let grtr = krates::semver::VersionReq::parse(">0.2.8").unwrap();
        let mut iter = krates.search_matches("winapi", grtr);

        let win38 = iter.next().unwrap();
        assert_eq!(win38.krate.name, "winapi");
        assert_eq!(
            win38.krate.version,
            krates::semver::Version::parse("0.3.9").unwrap()
        );

        assert!(iter.next().is_none());
    }

    {
        let none = krates::semver::VersionReq::parse("=0.4").unwrap();
        let mut iter = krates.search_matches("winapi", none);

        assert!(iter.next().is_none());
    }
}

#[test]
fn direct_dependents() {
    let mut kb = krates::Builder::new();
    kb.include_targets(std::iter::once((
        krates::cfg_expr::targets::get_builtin_target_by_triple("x86_64-unknown-linux-gnu")
            .unwrap()
            .triple
            .clone(),
        vec![],
    )));

    let grafs = util::build("direct.json", kb).unwrap();

    let id = grafs
        .actual
        .krates()
        .find(|k| k.0.repr.starts_with("reqwest"))
        .unwrap();

    let mut ids: Vec<_> = grafs
        .actual
        .direct_dependents(grafs.actual.nid_for_kid(&id.0).unwrap())
        .into_iter()
        .map(|jid| jid.krate.0.repr.as_str())
        .collect();

    ids.sort();
    let dd = ids.join("\n");

    insta::assert_snapshot!(dd);
}

#[test]
fn direct_dependencies() {
    let mut kb = krates::Builder::new();
    kb.include_targets(std::iter::once((
        krates::cfg_expr::targets::get_builtin_target_by_triple("x86_64-unknown-linux-gnu")
            .unwrap()
            .triple
            .clone(),
        vec![],
    )));

    let grafs = util::build("direct.json", kb).unwrap();

    let id = grafs
        .actual
        .krates()
        .find(|k| k.0.repr.starts_with("reqwest"))
        .unwrap();

    let mut ids: Vec<_> = grafs
        .actual
        .direct_dependencies(grafs.actual.nid_for_kid(&id.0).unwrap())
        .into_iter()
        .map(|jid| jid.krate.0.repr.as_str())
        .collect();

    ids.sort();
    let dd = ids.join("\n");

    insta::assert_snapshot!(dd);
}

#[test]
#[cfg(feature = "with-crates-index")]
fn bug_repro() {
    let mut kb = krates::Builder::new();
    kb.with_crates_io_index(None, krates::index::IndexKind::Sparse)
        .unwrap();

    let grafs = util::build("bug.json", kb).unwrap();

    insta::assert_snapshot!(grafs.dotgraph());
}
