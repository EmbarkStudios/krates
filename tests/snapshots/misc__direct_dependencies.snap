---
source: tests/misc.rs
expression: dd
---
base64 0.13.1 (registry+https://github.com/rust-lang/crates.io-index)
bytes 1.2.1 (registry+https://github.com/rust-lang/crates.io-index)
encoding_rs 0.8.31 (registry+https://github.com/rust-lang/crates.io-index)
futures-core 0.3.25 (registry+https://github.com/rust-lang/crates.io-index)
futures-util 0.3.25 (registry+https://github.com/rust-lang/crates.io-index)
h2 0.3.15 (registry+https://github.com/rust-lang/crates.io-index)
http 0.2.8 (registry+https://github.com/rust-lang/crates.io-index)
http-body 0.4.5 (registry+https://github.com/rust-lang/crates.io-index)
hyper 0.14.20 (registry+https://github.com/rust-lang/crates.io-index)
hyper-rustls 0.23.0 (registry+https://github.com/rust-lang/crates.io-index)
ipnet 2.5.0 (registry+https://github.com/rust-lang/crates.io-index)
libc 0.2.136 (registry+https://github.com/rust-lang/crates.io-index)
log 0.4.17 (registry+https://github.com/rust-lang/crates.io-index)
mime 0.3.16 (registry+https://github.com/rust-lang/crates.io-index)
mio 0.8.5 (registry+https://github.com/rust-lang/crates.io-index)
once_cell 1.15.0 (registry+https://github.com/rust-lang/crates.io-index)
percent-encoding 2.2.0 (registry+https://github.com/rust-lang/crates.io-index)
pin-project-lite 0.2.9 (registry+https://github.com/rust-lang/crates.io-index)
rustls 0.20.7 (registry+https://github.com/rust-lang/crates.io-index)
rustls-pemfile 1.0.1 (registry+https://github.com/rust-lang/crates.io-index)
serde 1.0.147 (registry+https://github.com/rust-lang/crates.io-index)
serde_json 1.0.87 (registry+https://github.com/rust-lang/crates.io-index)
serde_urlencoded 0.7.1 (registry+https://github.com/rust-lang/crates.io-index)
socket2 0.4.7 (registry+https://github.com/rust-lang/crates.io-index)
tokio 1.21.2 (registry+https://github.com/rust-lang/crates.io-index)
tokio-rustls 0.23.4 (registry+https://github.com/rust-lang/crates.io-index)
tower-service 0.3.2 (registry+https://github.com/rust-lang/crates.io-index)
url 2.3.1 (registry+https://github.com/rust-lang/crates.io-index)
webpki-roots 0.22.5 (registry+https://github.com/rust-lang/crates.io-index)
